File: hlm-backend/src/main/java/com/yem/hlm/backend/property/service/PropertyService.java
package com.yem.hlm.backend.property.service;

import com.yem.hlm.backend.immeuble.domain.Immeuble;
import com.yem.hlm.backend.immeuble.repo.ImmeubleRepository;
import com.yem.hlm.backend.immeuble.service.ImmeubleNotFoundException;
import com.yem.hlm.backend.project.service.ProjectActiveGuard;
import com.yem.hlm.backend.property.api.dto.BulkStatusRequest;
import com.yem.hlm.backend.property.api.dto.BulkStatusResult;
import com.yem.hlm.backend.property.api.dto.PropertyCreateRequest;
import com.yem.hlm.backend.property.api.dto.PropertyResponse;
import com.yem.hlm.backend.property.api.dto.PropertyStatusUpdateRequest;
import com.yem.hlm.backend.property.api.dto.PropertyUpdateRequest;
import com.yem.hlm.backend.property.domain.Property;
import com.yem.hlm.backend.property.domain.PropertyStatus;
import com.yem.hlm.backend.property.domain.PropertyType;
import com.yem.hlm.backend.property.repo.PropertyRepository;
import com.yem.hlm.backend.societe.CrossSocieteAccessException;
import com.yem.hlm.backend.societe.QuotaService;
import com.yem.hlm.backend.societe.SocieteContext;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

/**
 * Service for Property entity operations with type-specific validation.
 * <p>
 * Enforces validation rules based on PropertyType:
 * - VILLA: requires surface_area, land_area, bedrooms, bathrooms
 * - DUPLEX: requires surface_area, bedrooms, bathrooms, floors
 * - APPARTEMENT: requires surface_area, bedrooms, bathrooms, floor_number
 * - LOT: requires land_area, zoning, is_serviced; forbids bedrooms, bathrooms, building_year
 * - TERRAIN_VIERGE: requires land_area; forbids bedrooms, bathrooms, building_year, surface_area
 */
@Service
@Transactional(readOnly = true)
public class PropertyService {

    private final PropertyRepository propertyRepository;
    private final ProjectActiveGuard projectActiveGuard;
    private final PropertyCommercialWorkflowService propertyCommercialWorkflowService;
    private final ImmeubleRepository immeubleRepository;
    private final QuotaService quotaService;

    public PropertyService(PropertyRepository propertyRepository,
                           ProjectActiveGuard projectActiveGuard,
                           PropertyCommercialWorkflowService propertyCommercialWorkflowService,
                           ImmeubleRepository immeubleRepository,
                           QuotaService quotaService) {
        this.propertyRepository = propertyRepository;
        this.projectActiveGuard = projectActiveGuard;
        this.propertyCommercialWorkflowService = propertyCommercialWorkflowService;
        this.immeubleRepository = immeubleRepository;
        this.quotaService = quotaService;
    }

    @Transactional
    public PropertyResponse create(PropertyCreateRequest request) {
        UUID societeId = requireSocieteId();
        UUID userId = requireUserId();
        quotaService.enforceBienQuota(societeId);

        // Validate reference code uniqueness
        if (propertyRepository.existsBySocieteIdAndReferenceCode(societeId, request.referenceCode())) {
            throw new PropertyReferenceCodeExistsException(request.referenceCode());
        }

        // Type-specific validation
        validateTypeSpecificFields(request.type(), request);

        // Load project and assert it belongs to société and is ACTIVE
        var project = projectActiveGuard.requireActive(societeId, request.projectId());

        // Create entity
        var property = new Property(societeId, project, request.type(), userId);
        mapRequestToEntity(request, property);

        // Save
        property = propertyRepository.save(property);

        return PropertyResponse.from(property);
    }

    public PropertyResponse getById(UUID propertyId) {
        UUID societeId = requireSocieteId();

        var property = propertyRepository.findBySocieteIdAndIdAndDeletedAtIsNull(societeId, propertyId)
                .orElseThrow(() -> new PropertyNotFoundException(propertyId));

        return PropertyResponse.from(property);
    }

    public List<PropertyResponse> listAll() {
        return listAll(null, null, null, null);
    }

    public List<PropertyResponse> listAll(PropertyType type, PropertyStatus status) {
        return listAll(null, null, type, status);
    }

    public List<PropertyResponse> listAll(UUID projectId, UUID immeubleId,
                                           PropertyType type, PropertyStatus status) {
        UUID societeId = requireSocieteId();

        List<Property> properties = propertyRepository.findWithFilters(
                societeId, projectId, immeubleId, type, status);

        return properties.stream()
                .map(PropertyResponse::from)
                .collect(Collectors.toList());
    }

    @Transactional
    public PropertyResponse update(UUID propertyId, PropertyUpdateRequest request) {
        UUID societeId = requireSocieteId();
        UUID userId = requireUserId();

        var property = propertyRepository.findBySocieteIdAndId(societeId, propertyId)
                .orElseThrow(() -> new PropertyNotFoundException(propertyId));

        // Type-specific validation for the existing type
        validateTypeSpecificFieldsForUpdate(property.getType(), request);

        // Apply updates
        mapUpdateRequestToEntity(request, property);
        property.markUpdatedBy(userId);

        property = propertyRepository.save(property);

        return PropertyResponse.from(property);
    }

    @Transactional
    public void softDelete(UUID propertyId) {
        UUID societeId = requireSocieteId();

        var property = propertyRepository.findBySocieteIdAndId(societeId, propertyId)
                .orElseThrow(() -> new PropertyNotFoundException(propertyId));

        property.softDelete();
        propertyRepository.save(property);
    }

    /**
     * Changes the editorial status of a property.
     * <p>
     * Only non-commercial statuses may be set here: DRAFT, ACTIVE, WITHDRAWN, ARCHIVED.
     * RESERVED and SOLD are exclusively managed by the commercial workflow
     * (reservation/deposit/contract services) and are rejected with
     * {@link InvalidPropertyStatusTransitionException}.
     */
    @Transactional
    public PropertyResponse updateEditorialStatus(UUID propertyId, PropertyStatusUpdateRequest req) {
        PropertyStatus requested = req.status();
        if (requested == PropertyStatus.RESERVED || requested == PropertyStatus.SOLD) {
            throw new InvalidPropertyStatusTransitionException(
                    "Status " + requested + " can only be set by the commercial workflow " +
                    "(reservation / deposit / contract). Use the reservation or contract endpoints.");
        }

        UUID societeId = requireSocieteId();
        UUID userId = requireUserId();

        var property = propertyRepository.findBySocieteIdAndId(societeId, propertyId)
                .orElseThrow(() -> new PropertyNotFoundException(propertyId));

        // Guard: RESERVED and SOLD properties are under commercial lock —
        // do not allow editorial re-classification while a commercial event is active.
        if (property.getStatus() == PropertyStatus.RESERVED
                || property.getStatus() == PropertyStatus.SOLD) {
            throw new InvalidPropertyStatusTransitionException(
                    "Property is currently " + property.getStatus()
                    + " — cancel the reservation or contract before changing the editorial status.");
        }

        property.setStatus(requested);
        property.markUpdatedBy(userId);
        property = propertyRepository.save(property);
        return PropertyResponse.from(property);
    }

    /**
     * Applies a single editorial status to multiple properties in one transaction.
     * Properties that are RESERVED or SOLD, or not found in the société, are silently skipped
     * (returned in {@code skipped}) rather than failing the whole batch.
     * RESERVED and SOLD may not be set here — use the commercial workflow endpoints.
     */
    @Transactional
    public BulkStatusResult bulkUpdateEditorialStatus(BulkStatusRequest req) {
        if (req.status() == PropertyStatus.RESERVED || req.status() == PropertyStatus.SOLD) {
            throw new InvalidPropertyStatusTransitionException(
                    "Status " + req.status() + " cannot be set via the bulk endpoint — " +
                    "use the commercial workflow (reservation/contract).");
        }

        UUID societeId = requireSocieteId();
        UUID userId    = requireUserId();
        int updated = 0, skipped = 0;

        for (UUID id : req.ids()) {
            var opt = propertyRepository.findBySocieteIdAndId(societeId, id);
            if (opt.isEmpty()) { skipped++; continue; }

            var property = opt.get();
            if (property.getStatus() == PropertyStatus.RESERVED
                    || property.getStatus() == PropertyStatus.SOLD) {
                skipped++;
                continue;
            }

            property.setStatus(req.status());
            property.markUpdatedBy(userId);
            propertyRepository.save(property);
            updated++;
        }

        return new BulkStatusResult(updated, skipped);
    }

    @Transactional
    public void markAsReserved(UUID propertyId) {
        UUID societeId = requireSocieteId();

        var property = propertyRepository.findBySocieteIdAndId(societeId, propertyId)
                .orElseThrow(() -> new PropertyNotFoundException(propertyId));

        propertyCommercialWorkflowService.reserve(property, LocalDateTime.now());
    }

    @Transactional
    public void markAsSold(UUID propertyId, LocalDateTime soldAt) {
        UUID societeId = requireSocieteId();

        var property = propertyRepository.findBySocieteIdAndId(societeId, propertyId)
                .orElseThrow(() -> new PropertyNotFoundException(propertyId));

        propertyCommercialWorkflowService.sell(property, soldAt);
    }

    // ===== Type-Specific Validation Logic =====

    private void validateTypeSpecificFields(PropertyType type, PropertyCreateRequest req) {
        switch (type) {
            case VILLA -> {
                if (req.surfaceAreaSqm() == null) throw new InvalidPropertyTypeException("Surface area required for VILLA");
                if (req.landAreaSqm() == null) throw new InvalidPropertyTypeException("Land area required for VILLA");
                if (req.bedrooms() == null) throw new InvalidPropertyTypeException("Bedrooms required for VILLA");
                if (req.bathrooms() == null) throw new InvalidPropertyTypeException("Bathrooms required for VILLA");
            }
            case DUPLEX -> {
                if (req.surfaceAreaSqm() == null) throw new InvalidPropertyTypeException("Surface area required for DUPLEX");
                if (req.bedrooms() == null) throw new InvalidPropertyTypeException("Bedrooms required for DUPLEX");
                if (req.bathrooms() == null) throw new InvalidPropertyTypeException("Bathrooms required for DUPLEX");
                if (req.floors() == null) throw new InvalidPropertyTypeException("Floors required for DUPLEX");
            }
            case APPARTEMENT, T2, T3 -> {
                if (req.surfaceAreaSqm() == null) throw new InvalidPropertyTypeException("Surface area required for " + type);
                if (req.bedrooms() == null) throw new InvalidPropertyTypeException("Bedrooms required for " + type);
                if (req.bathrooms() == null) throw new InvalidPropertyTypeException("Bathrooms required for " + type);
                if (req.floorNumber() == null) throw new InvalidPropertyTypeException("Floor number required for " + type);
            }
            case STUDIO -> {
                if (req.surfaceAreaSqm() == null) throw new InvalidPropertyTypeException("Surface area required for STUDIO");
                if (req.floorNumber() == null) throw new InvalidPropertyTypeException("Floor number required for STUDIO");
            }
            case COMMERCE -> {
                if (req.surfaceAreaSqm() == null) throw new InvalidPropertyTypeException("Surface area required for COMMERCE");
            }
            case LOT -> {
                if (req.landAreaSqm() == null) throw new InvalidPropertyTypeException("Land area required for LOT");
                if (req.zoning() == null || req.zoning().isBlank()) throw new InvalidPropertyTypeException("Zoning required for LOT");
                if (req.isServiced() == null) throw new InvalidPropertyTypeException("is_serviced required for LOT");
                if (req.bedrooms() != null || req.bathrooms() != null || req.buildingYear() != null) {
                    throw new InvalidPropertyTypeException("Bedrooms/bathrooms/building_year not applicable to LOT");
                }
            }
            case TERRAIN_VIERGE -> {
                if (req.landAreaSqm() == null) throw new InvalidPropertyTypeException("Land area required for TERRAIN_VIERGE");
                if (req.bedrooms() != null || req.bathrooms() != null || req.buildingYear() != null || req.surfaceAreaSqm() != null) {
                    throw new InvalidPropertyTypeException("Bedrooms/bathrooms/building_year/surface_area not applicable to TERRAIN_VIERGE");
                }
            }
            case PARKING -> { /* no type-specific required fields — reference code and price are sufficient */ }
        }
    }

    private void validateTypeSpecificFieldsForUpdate(PropertyType type, PropertyUpdateRequest req) {
        switch (type) {
            case LOT -> {
                if (req.bedrooms() != null || req.bathrooms() != null || req.buildingYear() != null) {
                    throw new InvalidPropertyTypeException("Bedrooms/bathrooms/building_year not applicable to LOT");
                }
            }
            case TERRAIN_VIERGE -> {
                if (req.bedrooms() != null || req.bathrooms() != null || req.buildingYear() != null || req.surfaceAreaSqm() != null) {
                    throw new InvalidPropertyTypeException("Bedrooms/bathrooms/building_year/surface_area not applicable to TERRAIN_VIERGE");
                }
            }
            default -> { /* PARKING and other types: partial updates on physical fields are allowed */ }
        }
    }

    private void mapRequestToEntity(PropertyCreateRequest req, Property property) {
        property.setReferenceCode(req.referenceCode());
        property.setTitle(req.title());
        property.setDescription(req.description());
        property.setNotes(req.notes());
        property.setPrice(req.price());
        property.setCurrency(req.currency());
        property.setCommissionRate(req.commissionRate());
        property.setEstimatedValue(req.estimatedValue());
        property.setAddress(req.address());
        property.setCity(req.city());
        property.setRegion(req.region());
        property.setPostalCode(req.postalCode());
        property.setLatitude(req.latitude());
        property.setLongitude(req.longitude());
        property.setTitleDeedNumber(req.titleDeedNumber());
        property.setCadastralReference(req.cadastralReference());
        property.setOwnerName(req.ownerName());
        property.setLegalStatus(req.legalStatus());
        property.setSurfaceAreaSqm(req.surfaceAreaSqm());
        property.setLandAreaSqm(req.landAreaSqm());
        property.setBedrooms(req.bedrooms());
        property.setBathrooms(req.bathrooms());
        property.setFloors(req.floors());
        property.setParkingSpaces(req.parkingSpaces());
        property.setHasGarden(req.hasGarden());
        property.setHasPool(req.hasPool());
        property.setBuildingYear(req.buildingYear());
        property.setFloorNumber(req.floorNumber());
        property.setZoning(req.zoning());
        property.setIsServiced(req.isServiced());
        property.setListedForSale(req.listedForSale() != null && req.listedForSale());
        property.setBuildingName(req.buildingName());
        if (req.immeubleId() != null) {
            property.setImmeuble(loadImmeubleForProject(req.immeubleId(), property.getProject().getId()));
        }
    }

    private void mapUpdateRequestToEntity(PropertyUpdateRequest req, Property property) {
        if (req.title() != null) property.setTitle(req.title());
        if (req.description() != null) property.setDescription(req.description());
        if (req.notes() != null) property.setNotes(req.notes());
        if (req.price() != null) property.setPrice(req.price());
        if (req.address() != null) property.setAddress(req.address());
        if (req.city() != null) property.setCity(req.city());
        if (req.region() != null) property.setRegion(req.region());
        if (req.postalCode() != null) property.setPostalCode(req.postalCode());
        if (req.legalStatus() != null) property.setLegalStatus(req.legalStatus());
        if (req.surfaceAreaSqm() != null) property.setSurfaceAreaSqm(req.surfaceAreaSqm());
        if (req.landAreaSqm() != null) property.setLandAreaSqm(req.landAreaSqm());
        if (req.bedrooms() != null) property.setBedrooms(req.bedrooms());
        if (req.bathrooms() != null) property.setBathrooms(req.bathrooms());
        if (req.floors() != null) property.setFloors(req.floors());
        if (req.parkingSpaces() != null) property.setParkingSpaces(req.parkingSpaces());
        if (req.hasGarden() != null) property.setHasGarden(req.hasGarden());
        if (req.hasPool() != null) property.setHasPool(req.hasPool());
        if (req.buildingYear() != null) property.setBuildingYear(req.buildingYear());
        if (req.floorNumber() != null) property.setFloorNumber(req.floorNumber());
        if (req.zoning() != null) property.setZoning(req.zoning());
        if (req.isServiced() != null) property.setIsServiced(req.isServiced());
        if (req.listedForSale() != null) property.setListedForSale(req.listedForSale());
        UUID previousProjectId = property.getProject().getId();
        boolean projectChanged = false;
        if (req.projectId() != null) {
            UUID societeId = requireSocieteId();
            var project = projectActiveGuard.requireActive(societeId, req.projectId());
            property.setProject(project);
            projectChanged = !project.getId().equals(previousProjectId);
        }
        if (req.buildingName() != null) property.setBuildingName(req.buildingName());
        if (req.immeubleId() != null) {
            property.setImmeuble(loadImmeubleForProject(req.immeubleId(), property.getProject().getId()));
        } else if (projectChanged) {
            property.setImmeuble(null);
        }
    }

    private Immeuble loadImmeubleForProject(UUID immeubleId, UUID projectId) {
        UUID societeId = requireSocieteId();
        Immeuble immeuble = immeubleRepository.findBySocieteIdAndId(societeId, immeubleId)
                .orElseThrow(() -> new ImmeubleNotFoundException(immeubleId));

        if (!immeuble.getProject().getId().equals(projectId)) {
            throw new ImmeubleProjectMismatchException(immeubleId, projectId);
        }

        return immeuble;
    }

    private UUID requireSocieteId() {
        UUID id = SocieteContext.getSocieteId();
        if (id == null) throw new CrossSocieteAccessException("Missing société context");
        return id;
    }

    private UUID requireUserId() {
        UUID id = SocieteContext.getUserId();
        if (id == null) throw new CrossSocieteAccessException("Missing user context");
        return id;
    }
}

File: hlm-frontend/src/app/features/properties/property.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { environment } from '../../../environments/environment';
import { ImportResult, Property, PropertyMedia } from '../../core/models/property.model';
import { PagedResult } from '../../core/models/page-response.model';

/** Page size used by `list()` to fetch "all" matching rows for bounded callers (#023). */
const ALL_SIZE = 2000;

export interface CreatePropertyRequest {
  type: string;
  title: string;
  referenceCode: string;
  price: number;
  currency?: string | null;
  commissionRate?: number | null;
  estimatedValue?: number | null;
  address?: string | null;
  city?: string | null;
  region?: string | null;
  postalCode?: string | null;
  latitude?: number | null;
  longitude?: number | null;
  titleDeedNumber?: string | null;
  cadastralReference?: string | null;
  ownerName?: string | null;
  legalStatus?: string | null;
  surfaceAreaSqm?: number | null;
  landAreaSqm?: number | null;
  bedrooms?: number | null;
  bathrooms?: number | null;
  floors?: number | null;
  parkingSpaces?: number | null;
  hasGarden?: boolean | null;
  hasPool?: boolean | null;
  buildingYear?: number | null;
  floorNumber?: number | null;
  zoning?: string | null;
  isServiced?: boolean | null;
  description?: string | null;
  notes?: string | null;
  listedForSale?: boolean | null;
  projectId: string;
  immeubleId?: string | null;
  buildingName?: string | null;
}

export interface UpdatePropertyRequest {
  title?: string | null;
  description?: string | null;
  notes?: string | null;
  price?: number | null;
  address?: string | null;
  city?: string | null;
  region?: string | null;
  postalCode?: string | null;
  legalStatus?: string | null;
  surfaceAreaSqm?: number | null;
  landAreaSqm?: number | null;
  bedrooms?: number | null;
  bathrooms?: number | null;
  floors?: number | null;
  parkingSpaces?: number | null;
  hasGarden?: boolean | null;
  hasPool?: boolean | null;
  buildingYear?: number | null;
  floorNumber?: number | null;
  zoning?: string | null;
  isServiced?: boolean | null;
  listedForSale?: boolean | null;
  projectId?: string | null;
  immeubleId?: string | null;
  buildingName?: string | null;
}

export interface PropertyListParams {
  projectId?: string;
  immeubleId?: string;
  type?: string;
  status?: string;
}

/** VEFA commercial sheet. prixTtc is computed server-side; logementSocial is write-only (suggests TVA). */
export interface PropertyCommercial {
  propertyId?: string;
  etage?: number | null;
  orientation?: string | null;
  vue?: string | null;
  surfaceHabitable?: number | null;
  surfaceTerrasse?: number | null;
  surfaceCave?: number | null;
  surfaceParking?: number | null;
  parkingInclus?: boolean;
  caveIncluse?: boolean;
  prixHt?: number | null;
  tvaTaux?: number | null;
  prixTtc?: number | null;
  logementSocial?: boolean | null;
  penaliteRetardJournalier?: number | null;
  chargesCoproMensuelles?: number | null;
  planAppartementKey?: string | null;
}

@Injectable({ providedIn: 'root' })
export class PropertyService {
  private http = inject(HttpClient);
  private apiUrl = environment.apiUrl;

  /**
   * All matching properties (bounded callers: a building's units, a project's lots, pickers).
   * Backend is paginated (#023); request a large page and unwrap. Use `listPage()` for the
   * main properties page where real pagination is wanted.
   */
  list(params?: PropertyListParams): Observable<Property[]> {
    return this.http.get<PagedResult<Property>>(`${this.apiUrl}/api/properties`,
      { params: { ...this.filterParams(params), size: String(ALL_SIZE) } })
      .pipe(map(r => r.content));
  }

  /** Paginated filtered list for the main properties page (#023). */
  listPage(params: PropertyListParams | undefined, page: number, size: number):
      Observable<PagedResult<Property>> {
    return this.http.get<PagedResult<Property>>(`${this.apiUrl}/api/properties`,
      { params: { ...this.filterParams(params), page: String(page), size: String(size) } });
  }

  private filterParams(params?: PropertyListParams): Record<string, string> {
    const p: Record<string, string> = {};
    if (params?.projectId) p['projectId'] = params.projectId;
    if (params?.immeubleId) p['immeubleId'] = params.immeubleId;
    if (params?.type) p['type'] = params.type;
    if (params?.status) p['status'] = params.status;
    return p;
  }

  listMedia(propertyId: string): Observable<PropertyMedia[]> {
    return this.http.get<PropertyMedia[]>(`${this.apiUrl}/api/properties/${propertyId}/media`);
  }

  uploadMedia(propertyId: string, file: File): Observable<PropertyMedia> {
    const form = new FormData();
    form.append('file', file);
    return this.http.post<PropertyMedia>(`${this.apiUrl}/api/properties/${propertyId}/media`, form);
  }

  deleteMedia(mediaId: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/api/media/${mediaId}`);
  }

  downloadMediaUrl(mediaId: string): string {
    return `${this.apiUrl}/api/media/${mediaId}/download`;
  }

  importCsv(file: File): Observable<ImportResult> {
    const form = new FormData();
    form.append('file', file);
    return this.http.post<ImportResult>(`${this.apiUrl}/api/properties/import`, form);
  }

  /** Create a single property manually (ADMIN / MANAGER only). */
  create(req: CreatePropertyRequest): Observable<Property> {
    return this.http.post<Property>(`${this.apiUrl}/api/properties`, req);
  }

  /** Get a single property by ID. */
  getById(id: string): Observable<Property> {
    return this.http.get<Property>(`${this.apiUrl}/api/properties/${id}`);
  }

  /** Update an existing property (ADMIN / MANAGER only). */
  update(id: string, req: UpdatePropertyRequest): Observable<Property> {
    return this.http.put<Property>(`${this.apiUrl}/api/properties/${id}`, req);
  }

  // ── VEFA commercial sheet (HT/TVA/TTC) ──────────────────────────────────
  getCommercial(id: string): Observable<PropertyCommercial> {
    return this.http.get<PropertyCommercial>(`${this.apiUrl}/api/properties/${id}/commercial`);
  }

  updateCommercial(id: string, req: PropertyCommercial): Observable<PropertyCommercial> {
    return this.http.patch<PropertyCommercial>(`${this.apiUrl}/api/properties/${id}/commercial`, req);
  }

  /**
   * Change editorial status (ADMIN / MANAGER).
   * Only DRAFT / ACTIVE / WITHDRAWN / ARCHIVED are accepted.
   * RESERVED and SOLD are managed by the reservation/contract workflow.
   */
  setStatus(id: string, status: string): Observable<Property> {
    return this.http.patch<Property>(`${this.apiUrl}/api/properties/${id}/status`, { status });
  }

  /** Soft-delete a property (ADMIN only). */
  delete(id: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/api/properties/${id}`);
  }

  /**
   * Apply a single editorial status to multiple properties at once.
   * RESERVED/SOLD cannot be set here. Properties already in those statuses are skipped.
   */
  bulkSetStatus(ids: string[], status: string): Observable<{ updated: number; skipped: number }> {
    return this.http.patch<{ updated: number; skipped: number }>(
      `${this.apiUrl}/api/properties/bulk-status`,
      { ids, status }
    );
  }
}
